hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/update/server/SCMCRLUpdateHandler.java

```diff
@@ -0,0 +1,151 @@
+ * contributor license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.  The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.apache.hadoop.hdds.scm.update.server;
+
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.CRLUpdateResponse;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.UpdateResponse;
+import org.apache.hadoop.hdds.scm.update.client.CRLStore;
+import org.apache.hadoop.hdds.security.exception.SCMSecurityException;
+import org.apache.hadoop.hdds.security.x509.crl.CRLCodec;
+import org.apache.hadoop.hdds.security.x509.crl.CRLInfo;
+import org.apache.ratis.thirdparty.io.grpc.Status;
+import org.apache.ratis.thirdparty.io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Class handle the CRL client update and response.
+ */
+public class SCMCRLUpdateHandler implements SCMUpdateHandler {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SCMCRLUpdateHandler.class);
+  private final CRLStore crlStore;
+
+  private static final SCMUpdateServiceProtos.Type TYPE =
+      SCMUpdateServiceProtos.Type.CRLUpdate;
+
+  private final Map<UUID, CRLClientInfo> clients;
+
+  SCMCRLUpdateHandler(CRLStore crlStore) {
+    this.crlStore = crlStore;
+    clients = new ConcurrentHashMap<>();
+  }
+
+  public SCMUpdateServiceProtos.Type getType() {
+    return TYPE;
+  }
+
+  @Override
+  public void handleClientRequest(SCMUpdateServiceProtos.UpdateRequest request,
+      SCMUpdateClientInfo clientInfo) {
+    SCMUpdateServiceProtos.CRLUpdateRequest updateStatusRequest =
+        request.getCrlUpdateRequest();
+    long clientCrlId = updateStatusRequest.getReceivedCrlId();
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Client {} updateStatus \nclientCrlId {}  \npendingCrls {}",
+          clientInfo.getClientId(), clientCrlId,
+          updateStatusRequest.getPendingCrlIdsList().toString());
+    }
+
+    CRLClientInfo crlClientInfo;
+    if (!clients.containsKey(clientInfo.getClientId())) {
+      crlClientInfo = new CRLClientInfo(clientInfo);
+      clients.put(clientInfo.getClientId(), crlClientInfo);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **73 and 75**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **85**, **80**, and **82**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.